### PR TITLE
Warn if unsaved changes exist in assessment.

### PR DESCRIPTION
**Pull request by cahrens**
## [TNL-3870](https://openedx.atlassian.net/browse/TNL-3870)

This adds a warning if you try to browse away from a page where you have unsubmitted assessments. It covers peer assessments, staff override assessments, and staff full grade assessments (note you can have as many as 3 unsaved at once!). It builds on top of the infrastructure that warned if browsing away with an unsubmitted response.

To prevent the submission of a full grade assessment from wiping out a staff override assessment with unsaved changes, I had to do some refactoring of the full staff form. Because of this, now the "ungraded" and "checked out" counts will update when you check out an assessment for grading (which seems nice to me!).

### Sandbox
- [x] https://cahrens.sandbox.edx.org/courses/course-v1:ora+ora+ora/courseware/a4dfec19cf9b4a6fb5b18be6ccd9cecc/

### Testing
- [ ] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @andy-armstrong 
- [x] Code review: @efischer19 
- [ ] UI strings/error msgs review: @catong 
- [ ] Product review: @explorerleslie

FYI: @scottrish 

### Post-review
- [ ] Squash commits

### CODE REVIEW

File: openassessment/xblock/staff_area_mixin.py [hunk 1]
```diff
@@ -272,6 +272,32 @@ def render_staff_grade_form(self, data, suffix=''):  # pylint: disable=W0613
         except PeerAssessmentInternalError:
             return self.render_error(self._(u"Error getting staff grade information."))
 
+    @XBlock.handler
+    @require_course_staff("STUDENT_GRADE")
+    def render_staff_grade_counts(self, data, suffix=''):  # pylint: disable=W0613
```
Comment (efischer19) [inline openassessment/xblock/staff_area_mixin.py]: What is `suffix` used for here? It seems to be a common pattern so I doubt this needs to be fixed, I'm just curious.

File: openassessment/xblock/static/js/spec/lms/oa_peer.js [hunk 2]
```diff
@@ -115,4 +119,34 @@ describe("OpenAssessment.PeerView", function() {
...
+            $('input[type="radio"]', this).first().click();
+        });
+
+        expect(view.baseView.unsavedWarningEnabled()).toBe(true);
+
+        // When submitPeerAssessment is exected, the views will all re-render. However,
+        // as the test does not mock out the surrounding elements, the re-render
+        // of the peer assessment module will keep the original HTML intact (with selected
+        // options), causing the unsavedWarnings callback to be triggered again (after it is properly
+        // cleared during the submit operation). To avoid this, have the view re-render fail.
```
Comment (efischer19) [inline openassessment/xblock/static/js/spec/lms/oa_peer.js]: Good comment here :+1:

File: openassessment/xblock/static/js/spec/lms/oa_staff_area.js [hunk 3]
```diff
@@ -557,18 +588,75 @@ describe('OpenAssessment.StaffAreaView', function() {
...
+            // Fill in assessment and make sure the code re-renders the count form.
+            $assessment = getAssessment(staffArea.element, staffAreaTab);
+            fillAssessment($assessment);
+            expect(staffArea.baseView.unsavedWarningEnabled()).toBe(true);
+
+            submitAssessment(staffArea, staffAreaTab);
+            expect(staffArea.baseView.unsavedWarningEnabled()).toBe(false);
+        });
+
+        it('warns tracks unsubmitted assessments in multiple views', function() {
```
Comment (efischer19) [inline openassessment/xblock/static/js/spec/lms/oa_staff_area.js]: `warns tracks` seems odd to me, should this be just `tracks`?

**Comment by efischer19** [comment]
One minor note about the test name above, otherwise it looks good and performs like I'd expect on the sandbox. :ship: 

:+1:

File: openassessment/xblock/staff_area_mixin.py (hunk 1 above)
Comment (cahrens) [inline openassessment/xblock/staff_area_mixin.py]: Honestly, I don't know!

File: openassessment/xblock/staff_area_mixin.py [hunk 4]
```diff
@@ -272,6 +272,32 @@ def render_staff_grade_form(self, data, suffix=''):  # pylint: disable=W0613
...
+        Renders a form to show the number of ungraded and checked out assessments.
+
+        Must be course staff to render this view.
+        """
+        try:
+            student_item_dict = self.get_student_item_dict()
+            course_id = student_item_dict.get('course_id')
+            item_id = student_item_dict.get('item_id')
+
+            grading_stats = staff_api.get_staff_grading_statistics(course_id, item_id)
```
Comment (cahrens) [inline openassessment/xblock/staff_area_mixin.py]: I should create a helper method here (to share with code in get_staff_path_and_context).

File: openassessment/xblock/static/js/src/lms/oa_peer.js [hunk 5]
```diff
@@ -154,12 +162,29 @@ OpenAssessment.PeerView.prototype = {
...
+     * Used to set the unsaved changes warning dialog.
+     *
+     * @param {boolean} changesExist true if unsaved changes exist
+     */
+    assessmentRubricChanges: function(changesExist) {
+        if (changesExist) {
+            this.baseView.unsavedWarningEnabled(
+                true,
+                this.UNSAVED_WARNING_KEY,
+                gettext("If you leave this page without submitting your peer assessment, you'll lose any work you've done.") // jscs:ignore maximumLineLength
```
Comment (catong) [inline openassessment/xblock/static/js/src/lms/oa_peer.js]: "you'll" - > "you will" and "you've" -> "you have"

File: openassessment/xblock/static/js/src/lms/oa_response.js [hunk 6]
```diff
@@ -339,7 +310,11 @@ OpenAssessment.ResponseView.prototype = {
             this.saveEnabled(isNotBlank);
             this.previewEnabled(isNotBlank);
             this.saveStatus(gettext('This response has not been saved.'));
-            this.unsavedWarningEnabled(true);
+            this.baseView.unsavedWarningEnabled(
+                true,
+                this.UNSAVED_WARNING_KEY,
+                gettext("If you leave this page without saving or submitting your response, you'll lose any work you've done on the response.") // jscs:ignore maximumLineLength
```
Comment (catong) [inline openassessment/xblock/static/js/src/lms/oa_response.js]: "you'll" -> "you will" and
"you've" -> "you have"

File: openassessment/xblock/static/js/src/lms/oa_staff_area.js [hunk 7]
```diff
@@ -171,10 +186,27 @@
...
+         */
+        updateStaffGradeCounts: function() {
+            var view = this;
+            var $staffGradeTab = $('.openassessment__staff-grading', this.element);
+
+            view.server.staffGradeCounts().done(function(html) {
+                $staffGradeTab.find('.staff__grade__status').replaceWith(html);
+            }).fail(function() {
+                $staffGradeTab.find('.staff__grade__status').replaceWith(
+                    '<span class="staff__grade__status">gettext("Error getting number of ungraded assessments")</span>'
```
Comment (catong) [inline openassessment/xblock/static/js/src/lms/oa_staff_area.js]: Suggest add "the" for understanding, and change "assessments" to "responses": 
"Error getting the number of ungraded responses."

File: openassessment/xblock/static/js/src/lms/oa_staff_area.js [hunk 8]
```diff
@@ -380,6 +415,23 @@
...
+         *
+         * @param {string} key the unsaved changes key
+         * @param {boolean} changesExist true if unsaved changes exist
+         */
+        assessmentRubricChanges: function(key, changesExist) {
+            if (changesExist) {
+                this.baseView.unsavedWarningEnabled(
+                    true,
+                    key,
+                    gettext("If you leave this page without submitting your staff assessment, you'll lose any work you've done.") // jscs:ignore maximumLineLength
```
Comment (catong) [inline openassessment/xblock/static/js/src/lms/oa_staff_area.js]: "you'll" -. "you will" and "you've" -> "you have"

File: openassessment/xblock/static/js/src/oa_server.js [hunk 9]
```diff
@@ -137,6 +137,29 @@ if (typeof OpenAssessment.Server === "undefined" || !OpenAssessment.Server) {
...
+            return $.Deferred(function(defer) {
+                $.ajax({
+                    url: url,
+                    type: "POST",
+                    dataType: "html"
+                }).done(function(data) {
+                    defer.resolveWith(this, [data]);
+                }).fail(function() {
+                    defer.rejectWith(
+                        this, [gettext('The display of ungraded and checked out assessments could not be loaded.')]
```
Comment (catong) [inline openassessment/xblock/static/js/src/oa_server.js]: I think "assessments" -> "responses":
"The display of ungraded and checked out responses could not be loaded"

**Comment by cahrens** [comment]
@catong note that I'm showing the same message for both "staff override" unsaved changes and "full staff grading" unsaved changes. We can show different messages if we wish, but I wasn't sure how useful that would be.

**Comment by catong** [comment]
@cahrens I'm not sure which of my comments you're referring to? I didn't intend to provide edits that made the messages different. 
I changed "assessments" to "responses" for accuracy - staff are performing assessments, not grading assessments; it's "responses" that they're grading. I think we're consistent on that elsewhere.
Let me know if I've created some editing confusion!

**Comment by cahrens** [comment]
@catong I wasn't referring to anything you said-- just giving you a heads-up in case it wasn't obvious.

**Comment by catong** [comment]
@cahrens Ah, OK - thanks!

**Comment by explorerleslie** [comment]
:+1:

**Comment by cahrens** [comment]
Closing in favor of #826 against ora-staff-grading.

**Pull request closed without merging**